### PR TITLE
mkvs/proofs: Remove implicit internal leaf nodes from proofs

### PR DESCRIPTION
**Pull request by ptrus**
Internal MKVS nodes always implicitly contained the full leaf node in the proof. Instead it should be added as a regular child node in the proof. That way an unneded large value (e.g. value in a leaf node of an internal node on the path to the actual node we're proving) won't inflate the proofs size.

Change is backwards compatible, existing proofs are unmarshaled as V0.

TODO:
- [x] use the (commented out from existing test cases) V0 base64 proof encodings in a new test as V0 proof test vectors 
- [x] add a test case with non-nil internal leaf-node 
- [x] update rust part

### CODE REVIEW

File: go/storage/mkvs/syncer/proof.go [hunk 1]
```diff
@@ -113,7 +116,10 @@ func (b *ProofBuilder) Size() uint64 {
 
 // Build tries to build the proof.
 func (b *ProofBuilder) Build(ctx context.Context) (*Proof, error) {
-	var proof Proof
+	proof := Proof{
+		Versioned: cbor.NewVersioned(1),
```
Comment (kostko) [inline go/storage/mkvs/syncer/proof.go:120]: Add a `LatestProofVersion` constant.

Also probably we would need the builder to support both V0 and V1, so that old runtimes are not affected by this change.

File: go/storage/mkvs/node/node.go [hunk 2]
```diff
@@ -386,8 +386,25 @@ func (n *InternalNode) ExtractUnchecked() Node {
 }
 
 // CompactMarshalBinary encodes an internal node into binary form without
-// any hash pointers (e.g., for proofs).
+// any hash pointers and also doesn't include the leaf node (e.g., for proofs).
 func (n *InternalNode) CompactMarshalBinary() (data []byte, err error) {
```
Comment (kostko) [inline go/storage/mkvs/node/node.go:390]: We need to support serialization of V0 proofs for backwards compatibility.

**Review by kostko — COMMENTED** [review]
Also the read syncer interface should be changed so that it can request a specific version of the proof to make sure that old runtimes will continue to (implicitly) request V0 until they upgrade.

File: go/storage/mkvs/lookup.go [hunk 3]
```diff
@@ -48,12 +48,15 @@ func (t *tree) SyncGet(ctx context.Context, request *syncer.GetRequest) (*syncer
 	// Remember where the path from root to target node ends (will end).
 	t.cache.markPosition()
 
-	pb := syncer.NewProofBuilder(request.Tree.Root.Hash, request.Tree.Position)
+	pb, err := syncer.NewProofBuilderForVersion(request.Tree.Root.Hash, request.Tree.Position, 1)
```
Comment (peternose) [inline go/storage/mkvs/lookup.go:51]:
```suggestion
	pb := syncer.NewProofBuilder(request.Tree.Root.Hash, request.Tree.Position)
```

File: go/storage/mkvs/syncer/proof.go [hunk 4]
```diff
@@ -33,21 +44,37 @@ type proofNode struct {
 
 // ProofBuilder is a Merkle proof builder.
 type ProofBuilder struct {
-	root     hash.Hash
-	subtree  hash.Hash
-	included map[hash.Hash]*proofNode
-	size     uint64
+	proofVersion uint16
```
Comment (peternose) [inline go/storage/mkvs/syncer/proof.go:47]:
```suggestion
	proofVersion uint8
```

File: go/storage/mkvs/syncer/proof.go [hunk 5]
```diff
@@ -33,21 +44,37 @@ type proofNode struct {
...
+	root         hash.Hash
+	subtree      hash.Hash
+	included     map[hash.Hash]*proofNode
+	size         uint64
 }
 
 // NewProofBuilder creates a new Merkle proof builder for the given root.
 func NewProofBuilder(root, subtree hash.Hash) *ProofBuilder {
 	return &ProofBuilder{
```
Comment (peternose) [inline go/storage/mkvs/syncer/proof.go:56]:
```suggestion
	return NewProofBuilderForVersion(root, subtree, LatestProofVersion)
```

File: go/storage/mkvs/syncer/proof.go [hunk 6]
```diff
@@ -33,21 +44,37 @@ type proofNode struct {
...
+		subtree:      subtree,
+		included:     make(map[hash.Hash]*proofNode),
 	}
 }
 
+// NewProofBuilderForVersion creates a new Merkle proof builder for the given root
+// in a given proof version format.
+func NewProofBuilderForVersion(root, subtree hash.Hash, proofVersion uint16) (*ProofBuilder, error) {
+	if proofVersion < MinimumProofVersion || proofVersion > LatestProofVersion {
+		return nil, fmt.Errorf("proof builder: invalid proof version: %d", proofVersion)
```
Comment (peternose) [inline go/storage/mkvs/syncer/proof.go:68]:
```suggestion
		return nil, fmt.Errorf("proof: invalid proof version: %d", proofVersion)
```
I prefer not to return errors in constructors, if possible. As the programmer will select the version, we can maybe add a constructor for each version, e.g. `NewProofBuilder`, `NewProofBuilderV0`, `NewProofBuilderV1`, ... and use this constructor only as a helper (if needed).

File: go/storage/mkvs/checkpoint/chunk.go [hunk 7]
```diff
@@ -103,6 +103,7 @@ func restoreChunk(ctx context.Context, ndb db.NodeDB, chunk *ChunkMetadata, r io
 		p.Entries = append(p.Entries, entry)
 	}
 	p.UntrustedRoot = chunk.Root.Hash
+	p.V = 1
```
Comment (peternose) [inline go/storage/mkvs/checkpoint/chunk.go:106]:
```suggestion
	p.V = syncer.LatestProofVersion
```

File: go/storage/mkvs/node/node.go [hunk 8]
```diff
@@ -411,13 +407,47 @@ func (n *InternalNode) CompactMarshalBinary() (data []byte, err error) {
...
+	if n.LeafNode == nil {
+		leafNodeBinary = make([]byte, 1)
+		leafNodeBinary[0] = PrefixNilNode
+	} else {
+		if leafNodeBinary, err = n.LeafNode.Node.MarshalBinary(); err != nil {
+			return nil, fmt.Errorf("mkvs: failed to marshal leaf node: %w", err)
+		}
 	}
 
+	data = make([]byte, 1+DepthSize+len(n.Label)+len(leafNodeBinary))
```
[the remark this answers is not in the document]
Comment (ptrus) [reply go/storage/mkvs/node/node.go]: 👍 updated (also elsewhere)

File: go/storage/mkvs/lookup.go [hunk 9]
```diff
@@ -175,6 +180,12 @@ func (t *tree) doGet(
 		}
 
 		if opts.includeSiblings {
+			// Also fetch the leaf.
```
[the remark this answers is not in the document]
Comment (ptrus) [reply go/storage/mkvs/lookup.go]: Updated, used the first suggested version.

File: go/storage/mkvs/syncer/proof.go (hunk 6 above)
Comment (ptrus) [reply go/storage/mkvs/syncer/proof.go:68]: > ..As the programmer will select the version. 

That's not true in this case. The ReadSyncer request will contain the proof version (for backwards compatibility so that existing runtimes will get the current versions of proofs returned).

In places where the programmer selects the version the `NewProofBuilder` constructor is used (which also defaults to latest version).

File: go/storage/mkvs/syncer/proof.go (hunk 4 above)
Comment (ptrus) [reply go/storage/mkvs/syncer/proof.go:47]: Using `uint16` so that it is consistent with the version in the `Proof`s, which embeds our common `cbor.Versioned` helper struct that uses `uint16` for versions.

File: go/storage/mkvs/syncer/proof.go [hunk 10]
```diff
@@ -33,19 +56,50 @@ type proofNode struct {
...
+		proofVersion: LatestProofVersion,
+		root:         root,
+		subtree:      subtree,
+		included:     make(map[hash.Hash]*proofNode),
+	}
+}
+
+// NewProofBuilderV0 creates a new version 0 proof builder for the given root.
+func NewProofBuilderV0(root, subtree hash.Hash) *ProofBuilder {
+	return &ProofBuilder{
```
Comment (kostko) [inline go/storage/mkvs/syncer/proof.go:78]: Could also use `NewProofBuilderForVersion`.

File: go/storage/mkvs/syncer/proof_test.go [hunk 11]
```diff
@@ -35,12 +34,29 @@ func TestProofExtraNodes(t *testing.T) {
...
+
+	_, err = verifier.VerifyProof(context.Background(), rootHash, &proof)
+	require.Error(err, "proof with extra data should fail to validate")
 }
 
 func FuzzProof(f *testing.F) {
 	// Seed corpus.
-	rawProof, _ := base64.StdEncoding.DecodeString("omdlbnRyaWVzhUoBASQAa2V5IDACRgEBAQAAAlghAsFltYRhD4dAwHOdOmEigY1r02pJH6InhiibKlh9neYlWCECpsJnkjOnIgc4+yfvpsqCcIYHh5eld1hNMWTT7arAfHFYIQLhNTLWRbks1RBf52ulnlOTO+7D5EZNMYFzTx8U46sCnm51bnRydXN0ZWRfcm9vdFggWeZ8L9wIuOEN0Iu2uO/mFPzJZey4liX5fxf4fwcQRhM=")
-	f.Add(rawProof)
+	rawProofV0, _ := base64.StdEncoding.DecodeString("omdlbnRyaWVzhUoBASQAa2V5IDACRgEBAQAAAlghAsFltYRhD4dAwHOdOmEigY1r02pJH6InhiibKlh9neYlWCECpsJnkjOnIgc4+yfvpsqCcIYHh5eld1hNMWTT7arAfHFYIQLhNTLWRbks1RBf52ulnlOTO+7D5EZNMYFzTx8U46sCnm51bnRydXN0ZWRfcm9vdFggWeZ8L9wIuOEN0Iu2uO/mFPzJZey4liX5fxf4fwcQRhM=")
```
Comment (kostko) [inline go/storage/mkvs/syncer/proof_test.go:56]: Also re-run the fuzzer.

File: go/storage/mkvs/syncer/proof.go [hunk 12]
```diff
@@ -68,19 +120,43 @@ func (b *ProofBuilder) Include(n node.Node) {
 	// Node is available, serialize it.
 	var err error
 	var pn proofNode
-	pn.serialized, err = n.CompactMarshalBinary()
+	switch b.proofVersion {
+	case 0:
+		// In version 0, the leaf included is in the internal node.
```
Comment (kostko) [inline go/storage/mkvs/syncer/proof.go:125]:
```suggestion
		// In version 0, the leaf is included in the internal node.
```

File: runtime/src/storage/mkvs/sync/proof.rs [hunk 13]
```diff
@@ -450,49 +633,65 @@ uO/mFPzJZey4liX5fxf4fwcQRhM=",
...
-        let proof = tree
-            .get_proof(&keys[9])
-            .expect("get proof keys[9] works")
-            .expect("proof keys[9] exists");
-        assert_eq!(
-            base64::encode(cbor::to_vec(proof)),
-		    "omdlbnRyaWVzhUoBASQAa2V5IDACWCECJueKTLbwFMAiJitvfP3+tOruv3XChOjYSpH3U9/Xo/1GAQEDAIACWCECMMFu3slwotsl8hQsxQ/VPkrMtYMEsIrJAUH5PvSglANUAQAFAGtleSA5BwAAAHZhbHVlIDludW50cnVzdGVkX3Jvb3RYIFnmfC/cCLjhDdCLtrjv5hT8yWXsuJYl+X8X+H8HEEYT",
-        );
+        for tc in vec![
+            // Note: Tree::get_proof doesn't rust support version 0 proofs.
```
Comment (kostko) [inline runtime/src/storage/mkvs/sync/proof.rs:650]:
```suggestion
            // Note: Tree::get_proof doesn't support version 0 proofs.
```

**Review by kostko — APPROVED** [review]
